### ISSUE TITLE
Wrong descendant count after node moving

### ISSUE THREAD

**Issue by andrewshkovskii** [django-mptt/django-mptt#412]
We have an big tree (90k nodes).
Tree represents hierarchical structure of organisations in our country.
Sometimes we need to modify structure - move one or more organisations from one parent to new parent.
We do it using django transactions and mptt instance methods:
Our Node model look's like this:

```python
class TreeNode(mpttmodels.MPTTModel):
    TYPE_UNIT = 0
    TYPE_COMPANY = 1

    TYPE_CHOICES = (
        (TYPE_UNIT, "Administrative unit),
        (TYPE_COMPANY, "Company")
    )

    node_type = models.PositiveSmallIntegerField( choices=TYPE_CHOICES)
    node_element_id = models.IntegerField(db_index=True)
    parent = mpttmodels.TreeForeignKey(
        "self", null=True, blank=True, related_name="children", db_index=True)
    name = models.TextField(verbose_name="node name")
    administrative_unit_id = models.IntegerField(help_text="only for companies",
        null=True)
    is_authorized_organization = models.NullBooleanField(default=None)
```
Here is node move function:

```python
def node_move(node_obj, parent_company_node_obj):
    if node_obj.node_type == TreeNode.TYPE_COMPANY:
        with transaction.atomic():
            old_parent_node = node_obj.parent
            node_obj.move_to(parent_company_node_obj)
    else:
        raise NodeOperationError(node_obj)
```

So, for time to time, when you move a big bunch of nodes from one parent to another there 

The last time there was an error like this:
User moved 25 nodes from one parent to another and the new parent now has 51 children nodes.
But.. the new parent node instance methods was returning wrong calculations:

get_descendant_count method was returning 25
get_descendats().count() was returning 25 objects
children().count() - returning 51 objects (This one is true)

Call of manager rebuild() method solved this issue but how and why this can be possible if we are using transactions and correct methods ?

**Comment by cordery** [comment]
+1  I have this same issue.  I need to tally some aggregate information about a nodes children on save and not only is get_descendants often left in an empty state, but get_children is somehow not correctly populated either in the request where the parent is changed.  I actually monitor the node_moved signal and resave the target to try to get my  get_children based aggregates to tally correctly, but they frequently don't.  Quite baffled as to what is happening here.

**Comment by matthiask** [comment]
Yeah, you not only need transactions but also up-to-date MPTT attributes in your model instances before calling save() or anything on them. `Node.refresh_from_db()` before `Node.save()` helps if *anything* might have changed the MPTT attributes in the meantime.

**Comment by 21adrian1996** [comment]
Even if this issue seems to be quite old:

We have a model which has a lot of changes in the tree alignment.
So we remove, add, move a lot of objects in the tree and it regularly leads to broken trees.

In most cases we can check that the tree is broken and rebuild it if needed, but there are some cases which are hard to find,
as there seems to be no regularity in which cases the tree breaks.

Implementing the `refresh_from_db` call before the moves/updates seems to make it a bit more stable, but does not fully fix it.

Are there any other things we can do to improve the stability?
Is there some way to easily find broken trees?

**Comment by matthiask** [comment]
Duplicate of #257 — stale MPTT attributes on the instance at save time corrupt the tree. Call `instance.refresh_from_db()` before `save()` or `move_to()`. See #257 for the full discussion.
